File: leetcode/2342.数位和相等数对的最大和.go
func maximumSum(nums []int) int {
    //hash表能够到达O(1)
    hsh := map[int]int{}
    ans := -1
    for _, x := range nums {
        t := x
        tot := 0
        for t > 0 {
            tot += t % 10
            t /= 10
        }
        if hsh[tot] != 0 {
            ans = max(ans, x + hsh[tot])
        }
        hsh[tot] = max(hsh[tot], x)
    }
    return ans
}

File: leetcode/2342.数位和相等数对的最大和.cpp
class Solution {
public:
    int maximumSum(vector<int>& nums) {
    int n = nums.size();
    unordered_map<int, vector<int>> mp;
    for (int x : nums) {
        int sum = 0;
        int y = x;
        while (x > 0) {
            int a = x % 10;
            sum += a;
            x = x / 10;
        }
        mp[sum].push_back(y);
    }
    int mx = -1;
    for (auto x : mp) {
        vector<int> y = x.second;
        if (y.size() >= 2) {
            sort(y.begin(), y.end(), greater<int>());
            mx = max(mx, y[0] + y[1]);
        }
    }
    return mx;
}
};